[scroll-animations] Handle ViewTimeline subjects that are position:sticky

nmoucht opened the pull request:
#### 544a1a4c9970857dab48c9e052cab8b6ae091da1
<pre>
[scroll-animations] Handle ViewTimeline subjects that are position:sticky
<a href="https://bugs.webkit.org/show_bug.cgi?id=287451">https://bugs.webkit.org/show_bug.cgi?id=287451</a>
<a href="https://rdar.apple.com/144584129">rdar://144584129</a>

Reviewed by Simon Fraser.

To handle ViewTimeline subjects that are position:sticky, first ignore sticky offsets when getting the
subject offset from its scroll container. To compute the distance that a sticky element remains stationary,
it is necessary to get the difference of the sticky container&apos;s edge from its containing block&apos;s edge.
It is also necessary to compute the subjects location when the stickiness occurs, to ensure we can
properly adjust the range start and end, as well as for a specific animation-range. This will need a follow
up fix for targets larger than the scrollable container, which affects view-timeline-sticky-offscreen-6/7.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-none-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-view-functional-notation.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/sticky/view-timeline-sticky-offscreen-1-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/sticky/view-timeline-sticky-offscreen-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/sticky/view-timeline-sticky-offscreen-3-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/sticky/view-timeline-sticky-offscreen-6-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-sticky-block-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-sticky-block.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-sticky-inline-expected.txt:
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::StickinessAdjustmentData::computeStickinessAdjustmentData):
(WebCore::StickinessAdjustmentData::rangeStartAdjustment const):
(WebCore::StickinessAdjustmentData::rangeEndAdjustment const):
(WebCore::ViewTimeline::cacheCurrentTime):
(WebCore::ViewTimeline::stickyContainer const):
(WebCore::ViewTimeline::computeTimelineData const):
(WebCore::ViewTimeline::intervalForTimelineRangeName const):
(WebCore::operator&lt;&lt;):
* Source/WebCore/animation/ViewTimeline.h:

Canonical link: <a href="https://commits.webkit.org/290377@main">https://commits.webkit.org/290377@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e04d202f34a78f1a84b9b52c70b5e72015ee39d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89765 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44654 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94762 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40537 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91817 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9681 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17571 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69136 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26767 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92766 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7431 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81465 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49497 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7153 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35845 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39671 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77500 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36873 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96585 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16947 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12449 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78012 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17203 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77279 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77333 "Found 3 new API test failures: /TestWTF:WTF_EnumTraits.EnumNameSignedValues, /TestWTF:WTF_EnumTraits.EnumNameValid, /TestWTF:WTF_EnumTraits.EnumMinMax (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21767 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20357 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10129 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14107 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16960 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16701 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20159 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18483 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -198,6 +199,89 @@ AnimationTimelinesController* ViewTimeline::controller() const
     return nullptr;
 }
 
+StickinessAdjustmentData StickinessAdjustmentData::computeStickinessAdjustmentData(StickyPositionViewportConstraints& constraints, ScrollTimeline::ResolvedScrollDirection scrollDirection, float scrollContainerSize, float subjectSize, float subjectOffset)
```
`const StickyPositionViewportConstraints&`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -365,6 +459,23 @@ const RenderBox* ViewTimeline::sourceScrollerRenderer() const
+
+    auto scrollerRenderer = sourceScrollerRenderer();
+    while (renderer && renderer != scrollerRenderer) {
+        if (renderer->isStickilyPositioned())
```
What happens with nested stickies inside a scroller? Or a sticky inside a fixed position?

nmoucht replied:
Yeah both of these scenarios need some extra thought (and tests).

smfr commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -365,6 +459,23 @@ const RenderBox* ViewTimeline::sourceScrollerRenderer() const
+    while (renderer && renderer != scrollerRenderer) {
+        if (renderer->isStickilyPositioned())
+            return renderer.get();
+        renderer = renderer->parent();
```
Generally it's better to walk the containing block chain, since sticky forms a containing block. If `renderer` is a fixedpos, this code would be wrong.

nmoucht replied:
Fixed.

nmoucht committed:
[scroll-animations] Handle ViewTimeline subjects that are position:sticky
https://bugs.webkit.org/show_bug.cgi?id=287451
rdar://144584129

Reviewed by Simon Fraser.

To handle ViewTimeline subjects that are position:sticky, first ignore sticky offsets when getting the
subject offset from its scroll container. To compute the distance that a sticky element remains stationary,
it is necessary to get the difference of the sticky container's edge from its containing block's edge.
It is also necessary to compute the subjects location when the stickiness occurs, to ensure we can
properly adjust the range start and end, as well as for a specific animation-range. This will need a follow
up fix for targets larger than the scrollable container, which affects view-timeline-sticky-offscreen-6/7.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-none-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-view-functional-notation.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/sticky/view-timeline-sticky-offscreen-1-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/sticky/view-timeline-sticky-offscreen-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/sticky/view-timeline-sticky-offscreen-3-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/sticky/view-timeline-sticky-offscreen-6-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-sticky-block-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-sticky-block.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-sticky-inline-expected.txt:
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::StickinessAdjustmentData::computeStickinessAdjustmentData):
(WebCore::StickinessAdjustmentData::rangeStartAdjustment const):
(WebCore::StickinessAdjustmentData::rangeEndAdjustment const):
(WebCore::ViewTimeline::cacheCurrentTime):
(WebCore::ViewTimeline::stickyContainer const):
(WebCore::ViewTimeline::computeTimelineData const):
(WebCore::ViewTimeline::intervalForTimelineRangeName const):
(WebCore::operator<<):
* Source/WebCore/animation/ViewTimeline.h:

Canonical link: https://commits.webkit.org/290377@main
